heliopy/data/mms.py

```diff
@@ -7,131 +7,1781 @@
 """
 import os
 import pathlib
+import glob
+import datetime as dt
+import re
 import requests
-from tqdm.auto import tqdm
-from datetime import datetime, timedelta
+import tqdm.auto as tqdm
+import urllib
+import urllib3
+import pdb
+import io
+import csv
+import pathlib
+import warnings
 
+import heliopy
 from heliopy.data import util
-from heliopy import config
-
-data_dir = pathlib.Path(config['download_dir'])
+import sunpy.time
+import scipy.io
+from cdflib import epochs
+import numpy as np
+
+data_dir = pathlib.Path(heliopy.config['download_dir'])
+sdc_username = heliopy.config['mms_username']
+sdc_password = heliopy.config['mms_password']
 mms_dir = data_dir / 'mms'
 mms_url = 'https://lasp.colorado.edu/mms/sdc/public'
 remote_mms_dir = mms_url + '/data/'
 query_url = mms_url + '/files/api/v1/file_names/science'
 dl_url = mms_url + '/files/api/v1/download/science'
 
 
-def _validate_instrument(instrument):
-    allowed_instruments = ['afg', 'aspoc', 'dfg', 'dsp', 'edi',
-                           'edp', 'fgm', 'fpi', 'fields', 'scm', 'sdp', ]
-    if instrument not in allowed_instruments:
-        raise ValueError(
-            'Instrument {} not in list of allowed instruments: {}'.format(
-                instrument, allowed_instruments))
+class MMSDownloader(util.Downloader):
+    """
+    Download data from the Magnetospheric Multiscale (MMS) Science
+    Data Center (SDC). Data can be found at
+
+    https://lasp.colorado.edu/mms/sdc/
+
+    The following methods must be implemented by sub-classes:
+
+    - :meth:`Downloader.intervals()`: given a time interval, this
+      method should split the interval up into sub-intervals. Each of these
+      sub-intervals corresponds directly to a single file to download, store,
+      and read in.
+    - :meth:`Downloader.local_dir()`: given an interval, returns the
+      local directory in which the file is stored.
+    - :meth:`Downloader.fname()`: given an interval, returns the
+      local filename in which the file is stored.
+    - :meth:`Downloader.download()`: given an interval, download the data for
+      that interval.
+    - :meth:`Downloader.load_local_file()`: given an interval, load the local
+      file and return a :class:`pandas.DataFrame` object containing the data.
+    """
+    def __init__(self, sc=None, instr=None, mode=None, level=None,
+                 anc_product=None,
+                 data_type='science',
+                 data_root=None,
+                 dropbox_root=None,
+                 end_date=None,
+                 files=None,
+                 mirror_root=None,
+                 offline=False,
+                 optdesc=None,
+                 site='public',
+                 start_date=None,
+                 version=None):
+        """
+        Initialization method
+
+        Parameters
+        ----------
+        sc : str, list
```

Comment:
Why can these be strings or lists (I presume lists of strings?). Would be worth putting a note above the parameters section about these two options.